### PR TITLE
ADBDEV-6348: Add signal handlers to FtsProbe and WalReceiver

### PR DESCRIPTION
**Pull request by silent-observer**
Add signal handlers to FtsProbe and WalReceiver

Add signals handlers for SIGSEGV, SIGILL and SIGBUS to FtsProbe and
WalReceiver, similar to GPDB 6. The standard handler is used (the one that
logs a stacktrace) with corresponding process names.

Unlike GPDB 6, CdbProgramErrorHandler is not used for FtsProbe, because
FtsProbe does not use threads and therefore the standard handler is enough.

There are no tests added because testing for stacktraces appearing in logs is
complicated and unnecessary. To test this manually:
1. Run GPDB demo cluster with mirrors
2. Send a signal to FtsProbe or WalReceiver process with
`pkill -SIGSEGV -f ftsprobe` or `pkill -SIGSEGV -f walreceiver`.
4. Check GPDB logs for stacktraces

### CODE REVIEW

File: src/backend/replication/walreceiver.c [hunk 1]
```diff
@@ -131,6 +131,7 @@ static void WalRcvSigHupHandler(SIGNAL_ARGS);
 static void WalRcvSigUsr1Handler(SIGNAL_ARGS);
 static void WalRcvShutdownHandler(SIGNAL_ARGS);
 static void WalRcvQuickDieHandler(SIGNAL_ARGS);
+static void WalRcvCrashHandler(SIGNAL_ARGS);
```
Comment (RekGRpth) [inline src/backend/replication/walreceiver.c:134]: Why did you add a new function with a declaration in this file and without a declaration in another?

Comment (silent-observer) [reply src/backend/replication/walreceiver.c:134]: Because I placed the signal handler next to there handlers that were already present in the file. In WalReceiver they are below the place where the signals are registered and in FtsProbe the handlers are above.

**Comment by dkovalev1** [comment]
It looks good, but the question is: are there any tests possible? Or may be tests are exists already somewhere?

**Comment by silent-observer** [comment]
> Are there any tests possible?

Main result of this patch is enabling logging stacktraces if a segfault occurs inside one of these, how would you suggest testing that?

> Or may be tests are exists already somewhere?

I don't think so, the commits that added both of these handlers in GPDB 6 didn't have any tests, and I don't think there are any tests checking for signal handles in general since this is mostly for debug information.

**Comment by dkovalev1** [comment]
> > Are there any tests possible?
> 
> Main result of this patch is enabling logging stacktraces if a segfault occurs inside one of these, how would you suggest testing that?
1. Start the process
2. In the other shell run `kill -SIGILL <pid>` or `kill -SIGSEGV <pid>`
3. Check if traces are available
it's another question whether and how it should be automated, but we shall at least perform it and leave information how we did it.

**Comment by silent-observer** [comment]
Done: sending SIGSEGV to fts probe prints this to coordinator logs:
```
2024-09-25 08:26:42.923152 UTC,,,p46324,th0,,,,0,con3,,seg-1,,,,,"PANIC","XX000","Unexpected internal error: Master process received signal SIGSEGV",,,,,,,0,,,,"1    0x558f9c47b76d postgres StandardHandlerForSigillSigsegvSigbus_OnMainThread + 0x209
2    0x558f9c51d7ac postgres <symbol not found> + 0x9c51d7ac
3    0x558f9ca8f275 postgres <symbol not found> + 0x9ca8f275
4    0x7f45c5c42520 libc.so.6 <symbol not found> + 0xc5c42520
5    0x558f9c2532ff postgres WaitEventSetWait + 0x171
6    0x558f9c25267f postgres WaitLatchOrSocket + 0x1bc
7    0x558f9c2524c1 postgres WaitLatch + 0x3b
8    0x558f9c51e4b1 postgres <symbol not found> + 0x9c51e4b1
9    0x558f9c51d89c postgres FtsProbeMain + 0xad
"
```
And killing walreader prints this to mirror segment logs:
```
2024-09-25 08:28:05.955969 UTC,,,p47503,th0,,,,0,,,seg0,,,,,"PANIC","XX000","Unexpected internal error: Segment process received signal SIGSEGV",,,,,,,0,,,,"1    0x55e46e27276d postgres StandardHandlerForSigillSigsegvSigbus_OnMainThread + 0x209
2    0x55e46dff98ff postgres <symbol not found> + 0x6dff98ff
3    0x55e46e886275 postgres <symbol not found> + 0x6e886275
4    0x7febeb642520 libc.so.6 <symbol not found> + 0xeb642520
5    0x55e46e04a2ff postgres WaitEventSetWait + 0x171
6    0x55e46e04967f postgres WaitLatchOrSocket + 0x1bc
7    0x55e46dff8e89 postgres WalReceiverMain + 0xc55
"
```

**Comment by dkovalev1** [comment]
Great.
I don't see the need to add this to automated tests either, but could you reflect justification for it and this test scenario to the description?

**Review by dkovalev1 — APPROVED** [review]

**Comment by RekGRpth** [comment]
> * 
> 
> * Find FtsProbe or WalReceiver process with `ps -aux | grep fts` or
>   `ps -aux | grep walreceiver`
> 
> * Send a signal to it with `kill -SIGSEGV <pid>`

It can be done more simply
```sh
pkill -SIGSEGV -f ftsprobe
pkill -SIGSEGV -f walreceiver
```

**Comment by RekGRpth** [comment]
> sending SIGSEGV to fts probe prints this to coordinator logs:
> 
> ```
> 2024-09-25 08:26:42.923152 UTC,,,p46324,th0,,,,0,con3,,seg-1,,,,,"PANIC","XX000","Unexpected internal error: Master process received signal SIGSEGV",,,,,,,0,,,,"1    0x558f9c47b76d postgres StandardHandlerForSigillSigsegvSigbus_OnMainThread + 0x209
> 2    0x558f9c51d7ac postgres <symbol not found> + 0x9c51d7ac
> 3    0x558f9ca8f275 postgres <symbol not found> + 0x9ca8f275
> 4    0x7f45c5c42520 libc.so.6 <symbol not found> + 0xc5c42520
> 5    0x558f9c2532ff postgres WaitEventSetWait + 0x171
> 6    0x558f9c25267f postgres WaitLatchOrSocket + 0x1bc
> 7    0x558f9c2524c1 postgres WaitLatch + 0x3b
> 8    0x558f9c51e4b1 postgres <symbol not found> + 0x9c51e4b1
> 9    0x558f9c51d89c postgres FtsProbeMain + 0xad
> "
> ```

For some reason I get it like this:
```
2024-10-02 08:41:36.238587 +05,,,p9496,th0,,,,0,con3,,seg-1,,,,,"PANIC","XX000","Unexpected internal error: Master process received signal SIGSEGV",,,,,,,0,,,,"1    0xf8ae95 postgres StandardHandlerForSigillSigsegvSigbus_OnMainThread (elog.c:5095)
2    0x1019fdf postgres <symbol not found> (fts.c:102)
3    0x143938d postgres <symbol not found> (pqsignal.c:94)
"
```
although for this




> killing walreader prints this to mirror segment logs:
> 
> ```
> 2024-09-25 08:28:05.955969 UTC,,,p47503,th0,,,,0,,,seg0,,,,,"PANIC","XX000","Unexpected internal error: Segment process received signal SIGSEGV",,,,,,,0,,,,"1    0x55e46e27276d postgres StandardHandlerForSigillSigsegvSigbus_OnMainThread + 0x209
> 2    0x55e46dff98ff postgres <symbol not found> + 0x6dff98ff
> 3    0x55e46e886275 postgres <symbol not found> + 0x6e886275
> 4    0x7febeb642520 libc.so.6 <symbol not found> + 0xeb642520
> 5    0x55e46e04a2ff postgres WaitEventSetWait + 0x171
> 6    0x55e46e04967f postgres WaitLatchOrSocket + 0x1bc
> 7    0x55e46dff8e89 postgres WalReceiverMain + 0xc55
> "
> ```

everything is fine
```
2024-10-02 08:39:57.324976 +05,,,p8672,th0,,,,0,,,seg0,,,,,"PANIC","XX000","Unexpected internal error: Segment process received signal SIGSEGV",,,,,,,0,,,,"1    0xf8ae95 postgres StandardHandlerForSigillSigsegvSigbus_OnMainThread (elog.c:5095)
2    0xd56e04 postgres <symbol not found> (walreceiver.c:856)
3    0x143938d postgres <symbol not found> (pqsignal.c:94)
4    0x763fe5a12d20 libpthread.so.0 <symbol not found> + 0xe5a12d20
5    0xd9e4ab postgres WaitEventSetWait (latch.c:1111)
6    0xd9d9a9 postgres WaitLatchOrSocket (latch.c:408)
7    0xd564ef postgres WalReceiverMain (walreceiver.c:499)
"
```

**Comment by silent-observer** [comment]
> It can be done more simply

Ok, I changed the description.

> For some reason I get it like this

Probably related to stack unwinding issues, solved [here](https://github.com/arenadata/gpdb/pull/1057), which is currently blocked by this PR.

**Comment by RekGRpth** [comment]
> > For some reason I get it like this
> 
> Probably related to stack unwinding issues, solved [here](https://github.com/arenadata/gpdb/pull/1057), which is currently blocked by this PR.

Have you already provided a stack with a different patch?

**Comment by silent-observer** [comment]
> > > For some reason I get it like this
> > 
> > 
> > Probably related to stack unwinding issues, solved [here](https://github.com/arenadata/gpdb/pull/1057), which is currently blocked by this PR.
> 
> Have you already provided a stack with a different patch?

No, but it depends on build environment and specific version of dependencies, so it could easily work on one machine but not the other. This specific patch only guarantees that the SIGSEGV is logged at all, not the contents of the stacktrace.

**Review by RekGRpth — APPROVED** [review]

**Review by andr-sokolov — APPROVED** [review]
